uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/performance/DeviceMonitor.java

```diff
@@ -0,0 +1,248 @@
+     */
+    public static Double getCpuUsage() {
+        return getStats(PerformanceProfile.CPU, Double.class);
+    }
+
+    /**
+     * Get the current memory usage.
+     *
+     * @return the current memory usage in KB
+     */
+    public static Long getMemoryUsage() {
+        return getStats(PerformanceProfile.MEMORY, Long.class);
+    }
+
+    /**
+     * Get the current network info.
+     *
+     * @return the information regarding the network usage {@link TrafficInfo}
+     */
+    public static TrafficInfo getNetworkTrafficInfo() {
+        return getStats(PerformanceProfile.NETWORK, TrafficInfo.class);
+    }
+
+    /**
+     * Get the statistics of a particular performance profile.
+     *
+     * @param performanceProfile the performance profile
+     * @param classOfT           the class of the stats result
+     * @param <T>                the type of the stats retrieved for the performance profile
+     * @return the current data showing its usage.
+     */
+    public static <T> T getStats(@NonNull final PerformanceProfile performanceProfile, @NonNull final Class<T> classOfT) {
+        final PerformanceProfileMonitor<?> monitor = profilers.get(performanceProfile);
+
+        if (monitor == null) {
+            throw new IllegalArgumentException("The performanceProfile does not have a monitor.");
+        }
+
+        loadProcessInfo();
+
+        return classOfT.cast(monitor.getStats(processInfo));
+    }
+
+    /**
+     * Fetch the current process information.
+     *
+     * @return the current process information {@link ProcessInfo}
+     */
+    public static ProcessInfo getProcessInfo() {
+        return processInfo;
+    }
+
+    /**
+     * Updates or adds a new profiling monitor for a particular profile.
+     *
+     * @param profile  the profile to update {@link PerformanceProfile}
+     * @param profiler the profile monitor to use {@link PerformanceProfileMonitor}
+     * @param <T>      the type of the data that will be collected by the {@link PerformanceProfileMonitor}
+     */
+    public static <T> void setProfiler(@NonNull final PerformanceProfile profile, @NonNull final PerformanceProfileMonitor<T> profiler) {
+        profilers.put(profile, profiler);
+    }
+
+    /**
+     * Returns the current profile monitor tool that's being used by the profile.
+     *
+     * @param profile {@link PerformanceProfile}
+     * @return the profile monitor tool {@link PerformanceProfileMonitor}
+     */
+    public static PerformanceProfileMonitor<?> getProfiler(@NonNull final PerformanceProfile profile) {
+        return profilers.get(profile);
+    }
+
+    /**
+     * Fetches the current process information by executing several adb shell commands.
+     */
+    private static void loadProcessInfo() {
+        processInfo.setUid(getApplicationUid());
+
+        initializeCpuAndMemoryUsage();
+        initializeMemoryInfo();
+    }
+
+    /**
+     * Fetch the memory information by parsing output from /proc/meminfo
```

Comment:
I'm curious if there are other options Android provides to fetch these besides using shell commands?

Reply:
The only option was using `HardwarePropertiesManager` to get CPU information, but unfortunately this class is reserved for devices in Kiosk mode. It cannot be used by normal devices.